File: tools/mm2_event_lang/decompile.py
"""Decompile event.dat records into lifted AST."""

from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from decode_event import load_event_records, read_header  # noqa: E402

from .ast import EventFile, Location, RecordKind, Script, Stmt, StringDef, Trigger, TriggerCond
from .cfg_lift import lift_segment
from .parse_bytecode import (
    classify_record_kind,
    is_plain_text_segment,
    parse_location_record,
)
from .semantic_tables import (
    TRIGGER_COND_BY_BYTE,
    auto_script_name,
    auto_string_name,
    load_items_table,
)


def _trigger_cond(raw: int) -> TriggerCond:
    label = TRIGGER_COND_BY_BYTE.get(raw)
    if label == "always":
        return TriggerCond.ALWAYS
    if label == "enter":
        return TriggerCond.ENTER
    if label == "from north":
        return TriggerCond.FROM_NORTH
    if label == "dir special":
        return TriggerCond.DIR_SPECIAL
    if label == "any_direction":
        return TriggerCond.ANY_DIRECTION
    if label == "facing ns":
        return TriggerCond.FACING_NS
    if label == "enter special":
        return TriggerCond.ENTER_SPECIAL
    return TriggerCond.RAW


def decompile_location(
    blob: bytes,
    loc_id: int,
    items: list[str] | None = None,
) -> Location:
    items = items if items is not None else load_items_table()
    kind = RecordKind.UNKNOWN

    if not blob:
        return Location(id=loc_id, record_kind=kind, raw_blob=b"")

    loc_dict, segments_raw, segments_ops = parse_location_record(blob, loc_id)
    kind = classify_record_kind(loc_dict)
    strings_raw: list[str] = loc_dict["strings"]

    triplets_end = loc_dict["script_offset"] - 2
    raw_triplets = bytes(blob[:triplets_end]) if triplets_end > 0 else b""

    if kind == RecordKind.CASTLE_BLOB:
        return Location(
            id=loc_id,
            record_kind=kind,
            terminated=bool(loc_dict.get("terminated")),
            raw_blob=bytes(blob),
            raw_triplets=raw_triplets,
            comment="castle_blob — raw record preserved",
        )

    string_defs: list[StringDef] = []
    name_by_index: dict[int, str] = {}
    for i, text in enumerate(strings_raw):
        name = auto_string_name(i, text)
        name_by_index[i] = name
        string_defs.append(StringDef(name=name, text=text, index=i))

    triggers: list[Trigger] = []
    by_event: dict[int, list] = {}
    for pos, evt, cond_raw in loc_dict["triplets"]:
        y, x = (pos >> 4) & 0xF, pos & 0xF
        tc = _trigger_cond(cond_raw)
        script_name = auto_script_name(evt)
        triggers.append(
            Trigger(
                y=y,
                x=x,
                cond=tc,
                cond_raw=cond_raw,
                script_name=script_name,
                event_id=evt,
            )
        )
        by_event.setdefault(evt, []).append(triggers[-1])

    scripts: list[Script] = []
    for eid, seg_raw in enumerate(segments_raw):
        name = auto_script_name(eid)

        if is_plain_text_segment(seg_raw):
            text = seg_raw.decode("ascii", errors="replace").replace("@", "\n")
            scripts.append(
                Script(
                    name=name,
                    event_id=eid,
                    body=[Stmt.plain_text(text)],
                    is_plain_text=True,
                    raw_segment=bytes(seg_raw),
                )
            )
            continue

        ops = segments_ops[eid] if eid < len(segments_ops) else []
        body = lift_segment(ops, seg_raw, strings_raw, items)

        for t in by_event.get(eid, []):
            t.script_name = name

        for stmt in _walk_stmts(body):
            if stmt.kind in ("say", "service_title") and isinstance(stmt.data.get("string"), int):
                idx = stmt.data["string"]
                stmt.data["string"] = name_by_index.get(idx, f"s{idx}")

        scripts.append(
            Script(
                name=name,
                event_id=eid,
                body=body,
                is_plain_text=False,
                raw_segment=bytes(seg_raw),
            )
        )

    return Location(
        id=loc_id,
        record_kind=kind,
        terminated=bool(loc_dict.get("terminated")),
        triggers=triggers,
        strings=string_defs,
        scripts=scripts,
        raw_blob=bytes(blob),
        raw_triplets=raw_triplets,
    )


def _walk_stmts(stmts):
    from .ast import Stmt

    for s in stmts:
        yield s
        if s.kind == "if":
            yield from _walk_stmts(s.data.get("then") or [])
            yield from _walk_stmts(s.data.get("else") or [])
        elif s.kind == "unlifted":
            yield from _walk_stmts(s.data.get("body") or [])


def decompile_file(path: Path, items: list[str] | None = None) -> EventFile:
    data = path.read_bytes()
    header, records = load_event_records(data)
    locations = [decompile_location(rec, i, items) for i, rec in enumerate(records)]
    return EventFile(locations=locations, raw_records=list(records), header=header)

File: editor/src/eventlang/Decompile.cpp
#include "eventlang/Decompile.h"

#include "eventlang/BytecodeParse.h"
#include "eventlang/CfgLift.h"
#include "eventlang/Semantics.h"

#include <tuple>
#include <unordered_map>

namespace mm2::eventlang {
namespace {

void walkReplaceStringRefs(std::vector<Stmt>& stmts,
                           const std::unordered_map<int, std::string>& nameByIndex) {
    for (auto& s : stmts) {
        if (s.kind == "say" || s.kind == "service_title") {
            auto it = s.num.find("string");
            if (it != s.num.end()) {
                int idx = it->second;
                auto nit = nameByIndex.find(idx);
                s.str["string"] = nit != nameByIndex.end() ? nit->second : ("s" + std::to_string(idx));
                s.num.erase(it);
            }
        }
        if (s.kind == "if") {
            walkReplaceStringRefs(s.thenBody, nameByIndex);
            walkReplaceStringRefs(s.elseBody, nameByIndex);
        } else if (s.kind == "unlifted") {
            walkReplaceStringRefs(s.body, nameByIndex);
        }
    }
}

}  // namespace

bool loadEventRecords(const uint8_t* data, size_t len, EventFileAst& out) {
    out = EventFileAst{};
    constexpr size_t kHeader = 71 * 6;
    if (!data || len < kHeader) return false;
    out.header.resize(71);
    out.rawRecords.resize(71);
    for (int i = 0; i < 71; ++i) {
        const uint8_t* p = data + i * 6;
        uint32_t off = (static_cast<uint32_t>(p[0]) << 24) | (static_cast<uint32_t>(p[1]) << 16) |
                       (static_cast<uint32_t>(p[2]) << 8) | p[3];
        uint16_t length = static_cast<uint16_t>((p[4] << 8) | p[5]);
        out.header[i] = {off, length};
        if (off + length <= len) {
            out.rawRecords[i].assign(data + off, data + off + length);
        }
    }
    return true;
}

Location decompileLocation(const uint8_t* blob, size_t len, int locId) {
    Location loc;
    loc.id = locId;
    if (!blob || len == 0) {
        loc.recordKind = RecordKind::Unknown;
        return loc;
    }
    loc.rawBlob.assign(blob, blob + len);

    ParsedLocation parsed = parseLocationRecord(blob, len, locId);
    loc.recordKind = parsed.recordKind;
    loc.terminated = parsed.terminated;

    if (loc.recordKind == RecordKind::CastleBlob) {
        loc.comment = "castle_blob — raw record preserved";
        return loc;
    }
    // Overlay banks (60..70) decompile normally: scripts @2, strings @ LE anchor.

    std::unordered_map<int, std::string> nameByIndex;
    for (int i = 0; i < static_cast<int>(parsed.strings.size()); ++i) {
        StringDef sd;
        sd.name = autoStringName(i);
        sd.text = parsed.strings[i];
        sd.index = i;
        if (i < static_cast<int>(parsed.stringRaw.size())) sd.rawBytes = parsed.stringRaw[i];
        nameByIndex[i] = sd.name;
        loc.strings.push_back(std::move(sd));
    }

    std::unordered_map<int, std::vector<int>> triggerIdxByEvent;
    for (const auto& trip : parsed.triplets) {
        uint8_t pos = std::get<0>(trip);
        uint8_t evt = std::get<1>(trip);
        uint8_t condRaw = std::get<2>(trip);
        Trigger t;
        t.y = (pos >> 4) & 0xF;
        t.x = pos & 0xF;
        t.cond = triggerCondFromByte(condRaw);
        t.condRaw = condRaw;
        t.eventId = evt;
        t.scriptName = autoScriptName(evt);
        triggerIdxByEvent[evt].push_back(static_cast<int>(loc.triggers.size()));
        loc.triggers.push_back(std::move(t));
    }

    for (int eid = 0; eid < static_cast<int>(parsed.segmentsRaw.size()); ++eid) {
        const auto& segRaw = parsed.segmentsRaw[eid];
        Script sc;
        sc.name = autoScriptName(eid);
        sc.eventId = eid;
        sc.rawSegment = segRaw;

        if (looksLikeTextRecord(segRaw.data(), segRaw.size())) {
            std::string text;
            for (uint8_t b : segRaw) text += (b == 0x40) ? '\n' : static_cast<char>(b);
            sc.isPlainText = true;
            sc.body.push_back(Stmt::make("plain_text").set("text", text));
        } else {
            const auto& ops = parsed.segmentsOps[eid];
            sc.body = liftSegment(ops, segRaw.data(), segRaw.size());
            walkReplaceStringRefs(sc.body, nameByIndex);
        }

        for (int ti : triggerIdxByEvent[eid]) {
            if (ti >= 0 && ti < static_cast<int>(loc.triggers.size()))
                loc.triggers[ti].scriptName = sc.name;
        }
        loc.scripts.push_back(std::move(sc));
    }

    return loc;
}

EventFileAst decompileEventDat(const uint8_t* data, size_t len) {
    EventFileAst file;
    if (!loadEventRecords(data, len, file)) return file;
    file.locations.reserve(71);
    for (int i = 0; i < 71; ++i) {
        const auto& rec = file.rawRecords[i];
        file.locations.push_back(decompileLocation(rec.data(), rec.size(), i));
    }
    return file;
}

}  // namespace mm2::eventlang
